npm/design-system/src/library/hooks/useFuzzySearch.ts

```diff
@@ -0,0 +1,74 @@
+import Fuse from 'fuse.js'
+import { useEffect, useLayoutEffect, useMemo, useState } from 'react'
+
+/**
+ * Supports all Fuse options, with modified key definitions
+ */
+export interface FuzzySearchConfig<T> extends Omit<Fuse.IFuseOptions<T>, 'keys'> {
+  keys?: {
+    [Key in keyof T]?: true | {
+      weight: number
+    }
+  }
+}
+
+/**
+ * Provides fuzzy matching search over the supplied items, using weighted keys
+ * @param items The items to search
+ * @param searchInput A search string, provided if the hook is driven by external state
+ * @param config The configuration of how to perform searches, based on Fuse config options
+ *
+ * @see https://fusejs.io/api/options.html
+ */
+export const useFuzzySearch = <T>(items: T[], searchInput?: string, config?: FuzzySearchConfig<T>): {
+  searchInput: string
+  results: Fuse.FuseResult<T>[] | undefined
+  orderedResults: T[]
+  onSearch: (searchInput: string) => void
+} => {
+  const finalConfig = useMemo((): Fuse.IFuseOptions<T> | undefined => {
+    if (config?.keys === undefined) {
+      return config as Fuse.IFuseOptions<T>
+    }
+
+    return {
+      ...config,
+      keys: Object.keys(config.keys).map((key) => {
+        const keyEntry = config.keys?.[key as keyof T] as true | {
+          weight: number
+        } | undefined
+
+        if (keyEntry === undefined) {
```

Comment:
Any reason to prefer `=== undefined` over `(!keyEntry)`?

Reply:
In this scenario, yes, because `keyEntry` can be a boolean. Even though that boolean is true, that is a little too edgy for me, so I would rather explicitly check for the value I want here.